### PR TITLE
#9858 Fix for promotion of 100 percent with coupon

### LINKED ISSUE #9858: 100% Order percentage discount Coupon Based cannot be applied
**Sylius version affected**: 4.1.2

**Description**  
After creating a 100% Order percentage discount, i cannot apply it on an order

**Steps to reproduce**  
Create a 100% Order percentage discount Coupon Based.
Create a Cart, add a product, try to apply coupon on it.
Basket is not updated.

Thank you

### PR DESCRIPTION
**Pull request by laurent35240**
| Q               | A
| --------------- | -----
| Branch?         | 1.2
| Bug fix?        | yes
| New feature?    | no
| BC breaks?      | no
| Deprecations?   | no
| Related tickets | fixes #9858 
| License         | MIT

### CODE REVIEW

**Comment by Zales0123** [comment]
It's quite strange that this scenario passes... because it does not seem to work in dev environement 😕 

zrzut ekranu 2018-11-13 o 09 43 52
zrzut ekranu 2018-11-13 o 09 44 09
zrzut ekranu 2018-11-13 o 09 43 28
zrzut ekranu 2018-11-13 o 09 44 23

It needs some more investigation, there is totally something wrong happening here 😄 I've also confirmed (by going to website manually) that on `test` env it works 💥

**Comment by laurent35240** [comment]
@Zales0123 In fact the origin of the problem is the [PercentType field](https://symfony.com/doc/current/reference/forms/types/percent.html) in Symfony that will save a float for any value different than 100 and an int for 100.

If I save value 99% I got this in configuration field of table sylius_promotion_action
`a:1:{s:10:"percentage";d:0.99;}`
And with value 100%, I got this
`a:1:{s:10:"percentage";i:1;}`

I think the BDD test does not see it because it bypass the Symfony form by creating directly promotion action with a percentage set to 1 as a float.

If needed we can chat about a different/better way to fix this issue on slack

**Comment by laurent35240** [comment]
I found a different way to fix this issue cf https://github.com/laurent35240/Sylius/commit/5ba1a2d1042d7ce470eeae97d3f9b9918f6d499d

**Comment by Zales0123** [comment]
I've just checked that and it seems to work 💪 Could you please push the changes here so we could iterate over it and, hopefully, merge it? :)

**Comment by laurent35240** [comment]
Done

**Comment by Zales0123** [comment]
@laurent35240 the build failed, could you, please, take a look at it?

**Comment by laurent35240** [comment]
Sure, I will try to take a look at it this evening

**Comment by laurent35240** [comment]
@Zales0123 Build fixed ;-)

File: src/Sylius/Bundle/PromotionBundle/Form/Type/Action/PercentFloatType.php [hunk 1]
```diff
@@ -0,0 +1,16 @@
+<?php
+
+namespace Sylius\Bundle\PromotionBundle\Form\Type\Action;
+
+
+use Sylius\Bundle\PromotionBundle\Form\DataTransformer\PercentFloatToLocalizedStringTransformer;
+use Symfony\Component\Form\Extension\Core\Type\PercentType;
+use Symfony\Component\Form\FormBuilderInterface;
+
+class PercentFloatType extends PercentType
```
Comment (Zales0123) [inline src/Sylius/Bundle/PromotionBundle/Form/Type/Action/PercentFloatType.php]: I'm wondering, do we need this class? Maybe we can just call `$builder->get('percentage')->addViewTransformer(new PercentFloatToLocalizedStringTransformer());` in `PercentageDiscountConfigurationType`?

File: src/Sylius/Bundle/PromotionBundle/Form/DataTransformer/PercentFloatToLocalizedStringTransformer.php [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    /**
+     * Transforms between a percentage value into a float
+     *
+     * @param string $value Percentage value
+     *
+     * @return float Normalized value
+     *
+     * @throws TransformationFailedException if the given value is not a string or
+     *                                       if the value could not be transformed
+     */
```
Comment (Zales0123) [inline src/Sylius/Bundle/PromotionBundle/Form/DataTransformer/PercentFloatToLocalizedStringTransformer.php]: I think this docblock is redundant, the name of the class speaks for itself.

File: src/Sylius/Bundle/PromotionBundle/Form/DataTransformer/PercentFloatToLocalizedStringTransformer.php [hunk 3]
```diff
@@ -0,0 +1,28 @@
+<?php
+
+namespace Sylius\Bundle\PromotionBundle\Form\DataTransformer;
+
+
+use Symfony\Component\Form\Exception\TransformationFailedException;
+use Symfony\Component\Form\Extension\Core\DataTransformer\PercentToLocalizedStringTransformer;
+
+class PercentFloatToLocalizedStringTransformer extends PercentToLocalizedStringTransformer
```
Comment (Zales0123) [inline src/Sylius/Bundle/PromotionBundle/Form/DataTransformer/PercentFloatToLocalizedStringTransformer.php]: Should be final

File: src/Sylius/Bundle/PromotionBundle/Form/DataTransformer/PercentFloatToLocalizedStringTransformer.php [hunk 4]
```diff
@@ -0,0 +1,28 @@
+<?php
+
+namespace Sylius\Bundle\PromotionBundle\Form\DataTransformer;
+
+
```
Comment (Zales0123) [inline src/Sylius/Bundle/PromotionBundle/Form/DataTransformer/PercentFloatToLocalizedStringTransformer.php]: Doubled blank line

File: src/Sylius/Bundle/PromotionBundle/Form/DataTransformer/PercentFloatToLocalizedStringTransformer.php [hunk 5]
```diff
@@ -0,0 +1,28 @@
+<?php
+
```
Comment (Zales0123) [inline src/Sylius/Bundle/PromotionBundle/Form/DataTransformer/PercentFloatToLocalizedStringTransformer.php]: Missing license block and `declare(strict_types=1);` declaration

File: src/Sylius/Bundle/PromotionBundle/Form/DataTransformer/PercentFloatToLocalizedStringTransformer.php (hunk 2 above)
Comment (laurent35240) [reply src/Sylius/Bundle/PromotionBundle/Form/DataTransformer/PercentFloatToLocalizedStringTransformer.php]: I think it is good to keep it for at least the part with `@throws` in order to know which exception can be thrown by this method

File: src/Sylius/Bundle/PromotionBundle/Form/Type/Action/PercentFloatType.php (hunk 1 above)
Comment (laurent35240) [reply src/Sylius/Bundle/PromotionBundle/Form/Type/Action/PercentFloatType.php]: Yes i removed it and put instead
```php
 $builder->get('percentage')->resetViewTransformers()->addViewTransformer(new PercentFloatToLocalizedStringTransformer());```

**Comment by laurent35240** [comment]
HI

@Zales0123 Do you want other corrections on this PR?

**Comment by Zales0123** [comment]
Thanks, Laurent! :1st_place_medal:

**Pull request merged**